src/library_manager/llext_manager.c

```diff
@@ -71,177 +80,219 @@ static int llext_manager_align_unmap(void __sparse_cache *vma, size_t size)
-	dcache_writeback_region(va_base_rodata, st_rodata_size);
+	/* Copy writable data */
+	ret = llext_manager_load_data_from_storage(va_base_data, src_data,
+						   data_size, SYS_MM_MEM_PERM_RW);
+	if (ret < 0)
+		goto e_rodata;
+
+	memset((__sparse_force void *)bss_addr, 0, bss_size);
 
 	return 0;
 
+e_rodata:
+	llext_manager_align_unmap(va_base_rodata, rodata_size);
 e_text:
-	llext_manager_align_unmap(va_base_text, st_text_size);
+	llext_manager_align_unmap(va_base_text, text_size);
 
 	return ret;
 }
 
 static int llext_manager_unload_module(uint32_t module_id, const struct sof_man_module *mod)
 {
 	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
+	/* Executable code (.text) */
 	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	size_t st_text_size = ctx->segment_size[SOF_MAN_SEGMENT_TEXT];
+		ctx->segment[LIB_MANAGER_TEXT].addr;
+	size_t text_size = ctx->segment[LIB_MANAGER_TEXT].size;
+
+	/* Read-only data (.rodata, etc.) */
 	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	size_t st_rodata_size = ctx->segment_size[SOF_MAN_SEGMENT_RODATA];
+		ctx->segment[LIB_MANAGER_RODATA].addr;
+	size_t rodata_size = ctx->segment[LIB_MANAGER_RODATA].size;
+
+	/* Writable data (.data, .bss, etc.) */
+	void __sparse_cache *va_base_data = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_DATA].addr;
+	size_t data_size = ctx->segment[LIB_MANAGER_DATA].size;
 	int ret;
 
-	ret = llext_manager_align_unmap(va_base_text, st_text_size);
+	ret = llext_manager_align_unmap(va_base_text, text_size);
 	if (ret < 0)
 		return ret;
 
-	return llext_manager_align_unmap(va_base_rodata, st_rodata_size);
-}
-
-static void __sparse_cache *llext_manager_get_bss_address(uint32_t module_id,
-							  const struct sof_man_module *mod)
-{
-	return (void __sparse_cache *)mod->segment[SOF_MAN_SEGMENT_BSS].v_base_addr;
-}
-
-static int llext_manager_allocate_module_bss(uint32_t module_id,
-					     const struct sof_man_module *mod)
-{
-	/* FIXME: just map .bss together with .data and simply memset(.bss, 0) */
-	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
-	size_t bss_size = ctx->segment_size[SOF_MAN_SEGMENT_BSS];
-	void __sparse_cache *va_base = llext_manager_get_bss_address(module_id, mod);
-
-	/* Map bss memory and clear it. */
-	if (llext_manager_align_map(va_base, bss_size, SYS_MM_MEM_PERM_RW) < 0)
-		return -ENOMEM;
-
-	memset((__sparse_force void *)va_base, 0, bss_size);
-
-	return 0;
-}
-
-static int llext_manager_free_module_bss(uint32_t module_id,
-					 const struct sof_man_module *mod)
-{
-	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
-	size_t bss_size = ctx->segment_size[SOF_MAN_SEGMENT_BSS];
-	void __sparse_cache *va_base = llext_manager_get_bss_address(module_id, mod);
+	ret = llext_manager_align_unmap(va_base_data, data_size);
+	if (ret < 0)
+		return ret;
```

Comment:
Shouldn't we try to unmap the `rodata` even if unmapping the `data` fails?

Reply:
@softwarecki ok, this actually might make sense. Will update, thanks!